cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java

```diff
@@ -93,16 +98,28 @@ public DefaultStreamWriter(@Assisted("run") Id.Run run,
     this.runtimeUsageRegistry = runtimeUsageRegistry;
     this.authenticationContext = authenticationContext;
     this.authorizationEnabled = cConf.getBoolean(Constants.Security.Authorization.ENABLED);
+    this.retryStrategy = retryStrategy;
   }
 
   private URL getStreamURL(String stream) throws IOException {
     return getStreamURL(stream, false);
   }
 
   private URL getStreamURL(String stream, boolean batch) throws IOException {
-    Discoverable discoverable = endpointStrategy.pick(1, TimeUnit.SECONDS);
-    if (discoverable == null) {
-      throw new IOException("Stream Service Endpoint not found");
+    Discoverable discoverable;
+    try {
+      discoverable = Retries.supplyWithRetries(new Supplier<Discoverable>() {
+        @Override
+        public Discoverable get() {
+          Discoverable discoverable = endpointStrategy.pick();
+          if (discoverable == null) {
+            throw new ServiceUnavailableException(Constants.Service.STREAMS);
+          }
+          return discoverable;
+        }
+      }, retryStrategy);
+    } catch (InterruptedException e) {
+      throw new ServiceUnavailableException(Constants.Service.STREAMS);
```

Comment:
Why not catch RetriesExhaustedException and throw its cause instead? To the caller, this should still be a ServiceUnavailableException, or does it not matter?